### ISSUE TITLE
Unable to build WebDriverAgent for parallel executions with iOS Simulators 17.0 and Xcode 15 due to xcode build error 70

### ISSUE THREAD

**Issue by alexcc010** [appium/WebDriverAgent#806]
Hello guys,
Since i updated to Xcode 15, im not able to run iOS parallel test executions in a stable way, as i used to do with Xcode 14, due to **Xcode build exited with code 70**.

After some debugging, what happens is that as soon as I start the run, the two iPhone simulators launch and installs the app, both with a delay of around 30 seconds between each other, and so far without any problems.
However, when the WebDriverAgent build starts, **xcode's derived data path is constantly being cleared** since it ends up to be 2 simulators waiting for the WebDriverAgent xcode build, and therefore **'WebDriverAgentRunner_iphonesimulator17.0-arm64.xctestrun' ends up not existing because it is being deleted**. In previous versions of Xcode, while both WebDriverAgent project were building, the folder was not cleaned and the build ran for both simulators without any problems.
The only way i can now run parallel executions is to set a wait time of around 2 minutes between the simulators launch, which adds more execution time, and it's not clear to me how much time the WebDriverAgent build takes for each simulator, so that i can prevent in a efficient way the error code 70 and signal 'null', instead of just set a time.sleep().

My execution project is based on python and i handle the parallel execution by using python multiprocessing package.

- My specs are:
  - Appium version 2.2.1
  - Xcode version 15.0.1
  - xcuitest version 5.8.1
  - Two iPhone **Simulators** 17.0
  - MacOS Ventura 13.6.1

- iPhone Simulator Capabilities:
    - desired_caps = {
            'platformName': 'iOS',
            'platformVersion': self.device.version,
            'deviceName': self.device.name,
            'automationName': "XCUITest",
            'noReset': True,
            'fullReset': False,
            'app': self.__get_artifact(),
            'udid': self.device.uuid,
            'bundleId': myAppBundleId,
            'clearSystemFiles': True,
            'autoAcceptAlerts': False,
            'useNewWDA': False,
            'wdaLocalPort': self.wda_local_port,
            'newCommandTimeout': 4000,
            'language': "en"
        }

<img width="1208" alt="Screenshot 2023-11-06 at 11 05 19" src="https://github.com/appium/WebDriverAgent/assets/149820998/e1f2f7a1-9dca-4a51-a019-3fdd4497fa11">

**Comment by gforepsly** [comment]
We are facing very similar issues with the same stack.
Unable to start WebDriverAgent session because of xcodebuild failure: An unknown server-side error occurred while processing the command. Original error: Could not proxy command to the remote server. Original error: connect ECONNREFUSED [127.0.0.1:8668](http://127.0.0.1:8668/)

Our capabilities:
platformName: 'iOS',
            "appium:iosInstallPause": 5000,
            "appium:autoAcceptAlerts": true,
            "appium:autoDismissAlerts": true,
            "appium:deviceName": 'iPhone 14 Pro',
            "appium:language": 'en',
            "appium:platformVersion":  '17.0',
            "appium:automationName": 'XCUITest',
            "appium:app": 'pathToApp'
            "appium:bundleId": 'bundleIdWeUse',
            "appium:newCommandTimeout": 10000,
            "appium:wdaLaunchTimeout": 300000,
            "appium:wdaConnectionTimeout": 600000,
            "appium:wdaStartupRetries": 4,
            "appium:wdaStartupRetryInterval": 120000,
            "appium:noReset":  false,
            "appium:wdaLocalPort": 8668,
            "appium:useNewWDA": false,
            'appium:derivedDataPath': 'unique derived data path'

As soon as second or third simulator starts the tests, everything stops. Any luck or any solutions? We did everything according the documentation for parallel execution.

**Comment by KazuCocoa** [comment]
hm, it looks like Xcode/Simulator related environment itself is not so stable in the newer Xcode version...?

**Comment by gforepsly** [comment]
> hm, it looks like Xcode/Simulator related environment itself is not so stable in the newer Xcode version...?

Is there like a checklist for this? 
We are experiencing the same issues with Xcode14 and Xcode15. Xcode14 uses OS16.2 (older MAC) while Xcode15 OS17 (newer MAC) and we use that stack to run our tests. As soon as multiple simulators starts, often we are getting the same error:

Proxying [GET /status] to [GET http://127.0.0.1:8878/status] with no body
[XCUITestDriver@56ed (db392841)] connect ECONNREFUSED 127.0.0.1:8878

It works occasionally, then it doesn't, then it does... It's really frustrating.

**Comment by KazuCocoa** [comment]
No, it is more Xcode and simulator thing rather than Appium specific. So the Xcode tool chain themselves.

```
Proxying [GET /status] to [GET http://127.0.0.1:8878/status] with no body
[XCUITestDriver@56ed (db392841)] connect ECONNREFUSED 127.0.0.1:8878
```
Keep showing the message is expected. Appium keeps attempting to attach to the WDA via the URL. The error should happen until the connection between Appium running on the host machine and WDA running on the test device gets connected. for example xcodebuild is still building the WDA module and has not launched yet, the error must keep showing.

**Comment by gforepsly** [comment]
> Keep showing the message is expected. Appium keeps attempting to attach to the WDA via the URL. The error should happen until the connection between Appium running on the host machine and WDA running on the test device gets connected. for example xcodebuild is still building the WDA module and has not launched yet, the error must keep showing.

Would it help to kill Xcode and all SImulators + Appium, and to start them fresh every morning? 
What puzzles me is the next thing:
 - We are using xcodebuild to build .app of our application for the simulators.
 - We are pointing our simulators/tests to use that prebuilt .app for tests

Why does Xcode tries to build additional stuff? 

Our plan is to update XCUITEST to the latest version as soon as we refactor some deprecated methods in our code.

**Comment by KazuCocoa** [comment]
I have no good answer since I mostly don't use simulators for parallel running.
https://discuss.appium.io/ might help since some users would build similar configurations, I guess.

> Why does Xcode tries to build additional stuff?

Only Apple knows about this point. I wondered if `useXctestrunFile` would help https://appium.github.io/appium-xcuitest-driver/5.11/run-prebuilt-wda/#how-to-use-appiumusexctestrunfile-and-appiumbootstrappath-capabilities

**Comment by amedvedjev** [comment]
maybe it can it help to use pre-build WDA app instead of trying build it each test?

capability 
`String USE_PREBUILT_WDA_OPTION = "usePrebuiltWDA";`

**Comment by gforepsly** [comment]
> maybe it can it help to use pre-build WDA app instead of trying build it each test?
> 
> capability `String USE_PREBUILT_WDA_OPTION = "usePrebuiltWDA";`

Thx for the tip, but unfortunately, it works the same :( 
We can try using this option, maybe it will at least be more stable.

**Comment by WolfmosRioja** [comment]
> I have no good answer since I mostly don't use simulators for parallel running. https://discuss.appium.io/ might help since some users would build similar configurations, I guess.
> 
> > Why does Xcode try to build additional stuff?
> 
> Only Apple knows about this point. I wondered if `useXctestrunFile` would help https://appium.github.io/appium-xcuitest-driver/5.11/run-prebuilt-wda/#how-to-use-appiumusexctestrunfile-and-appiumbootstrappath-capabilities

for me it's not working with that, this is happening --> `[ERROR]   Run 1: An error occurred during initializing drivers  ---> org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: An unknown server-side error occurred while processing the command. Original error: Unable to launch WebDriverAgent because of xcodebuild failure: If you are using 'useXctestrunFile' capability then you need to have a xctestrun file (expected: 'URL/Build/Products/WebDriverAgentRunner_iphoneosnull-arm64.xctestrun'). Make sure you follow the tutorial at https://github.com/appium/appium-xcuitest-driver/blob/master/docs/real-device-config.md. Try to remove the WebDriverAgentRunner application from the device if it is installed and reboot the device. 
`
and here are talking about usePrebuiltWDA --> https://github.com/appium/appium/issues/18749#issuecomment-1652954536

**Comment by Jitu1888** [comment]
Any update ? @KazuCocoa

**Comment by KazuCocoa** [comment]
`WebDriverAgentRunner_iphoneosnull` This could be weird. The `null` might need to be the os version, so possibly capabilities/some logic missed the os version part in somewhere. Could you create a ticket in https://github.com/appium/appium with appium logs?

(Please create a new issue in the appium repository so that we could notice more well instead of this repo)

**Comment by tking16** [comment]
Can you please add the link to the ticket here too so we can track the status 🙂
